### DISCUSSION TITLE
python query of nominatim for foreign characters? (german addresses)

### DISCUSSION THREAD

**Discussion by tueboesen** [osm-search/Nominatim#3446] [Q&A]
I have set up a nominatim server for europe and it seems to work at first glance. I can query it and get reasonable replies.

The problem comes when I have strings with non ascii characters that I want to query.

My current setup looks like this:

```
from urllib.request import Request, urlopen
from urllib.parse import quote
def convert_streetaddress_to_coordinates(streetaddress,host='http://127.0.0.1:8080',service='search'):
    url = f"{host}/{service}?q={quote(streetaddress)}&limit=1"
    req = Request(url)
    response = urlopen(req)
    reponse_read= response.read().decode('utf-8')
    parsed_json = json.loads(reponse_read)
    try:
        lat = parsed_json[0]["lat"]
        lon = parsed_json[0]["lon"]
    except:
        return None
    return lat, lon

```

However for an address like:
`'PILGERSTRASSE 19,DE-45473  MÜLHEIM DÜMPTEN' `
I get the following:
`'http://10.150.0.30:8080/search?q=PILGERSTRASSE%2019%2CDE-45473%20%20M%C3%9CLHEIM%20D%C3%9CMPTEN&limit=1'`
Which returns nothing when passed to nominatim.

So how exactly do I pass such characters to nominatim?

**Comment by mtmail** [comment]
I don't think the URL escaping is an issue.

As far as I see the street is in [Altstadt II](https://www.openstreetmap.org/relation/4620179), not [Dümpten](https://www.openstreetmap.org/relation/4620180) suburb

The same query doesn't work on openstreetmap.org either 

![image](https://github.com/osm-search/Nominatim/assets/3727288/c9e9a0b9-a083-477c-8412-b9636efd738f)

**Reply by tueboesen** [reply]
Hmm I didn't consider that it would have trouble with the address as you suggest. However, I just tried to update the address to:
Pilgerstraße 19, 45473 Mülheim an der Ruhr, Germany
which works with both google maps and nominatim.openstreetmap.org

However when I convert that address I get the following:
`'http://10.150.0.30:8080/search?q=Pilgerstra%C3%9Fe%2B19%2C%2B45473%2BM%C3%BClheim%2Ban%2Bder%2BRuhr%2C%2BGermany&limit=1'`

which has no hits.

While I can see that the address on nominatim.openstreetmap.org is the following:
`https://nominatim.openstreetmap.org/ui/search.html?q=Pilgerstra%C3%9Fe+19%2C+45473+M%C3%BClheim+an+der+Ruhr%2C+Germany`

So how can I get a similar search address in python?

**Comment by lonvia** [comment]
Instead of writing your own code, have a look at GeoPy. The [Nominatim connector](https://geopy.readthedocs.io/en/latest/#nominatim) can be used with a custom `domain` parameter to point it at your own server.
